crates/matrix-sdk/src/room/reply.rs

```diff
@@ -0,0 +1,392 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Facilities to reply to existing events.
+
+use std::future::Future;
+
+use matrix_sdk_base::{deserialized_responses::TimelineEvent, SendOutsideWasm};
+use ruma::{
+    events::{
+        relation::Thread,
+        room::{
+            encrypted::Relation as EncryptedRelation,
+            message::{
+                AddMentions, ForwardThread, OriginalRoomMessageEvent, Relation, ReplyWithinThread,
+                RoomMessageEventContent, RoomMessageEventContentWithoutRelation,
+            },
+        },
+        AnyMessageLikeEventContent, AnySyncMessageLikeEvent, AnySyncTimelineEvent,
+        SyncMessageLikeEvent,
+    },
+    serde::Raw,
+    EventId, MilliSecondsSinceUnixEpoch, OwnedEventId, OwnedUserId, RoomId, UserId,
+};
+use serde::Deserialize;
+use thiserror::Error;
+use tracing::{error, instrument};
+
+use super::Room;
+
+/// Information needed to reply to an event.
+#[derive(Debug, Clone)]
+struct RepliedToInfo {
+    /// The event ID of the event to reply to.
+    event_id: OwnedEventId,
+    /// The sender of the event to reply to.
+    sender: OwnedUserId,
+    /// The timestamp of the event to reply to.
+    timestamp: MilliSecondsSinceUnixEpoch,
+    /// The content of the event to reply to.
+    content: ReplyContent,
+}
+
+/// The content of a reply.
+#[derive(Debug, Clone)]
+enum ReplyContent {
+    /// Content of a message event.
+    Message(RoomMessageEventContent),
+    /// Content of any other kind of event stored as raw JSON.
+    Raw(Raw<AnySyncTimelineEvent>),
+}
+
+/// Errors specific to unsupported replies.
+#[derive(Debug, Error)]
+pub enum ReplyError {
+    /// We couldn't fetch the remote event with /room/event.
+    #[error("Couldn't fetch the remote event: {0}")]
+    Fetch(Box<crate::Error>),
+    /// Redacted events whose JSON form isn't available cannot be replied to.
+    #[error("redacted events whose JSON form isn't available can't be replied")]
+    MissingJson,
+    /// The event to reply to could not be found.
+    #[error("event to reply to not found")]
+    MissingEvent,
+    /// The event to reply to could not be deserialized.
+    #[error("failed to deserialize event to reply to")]
+    FailedToDeserializeEvent,
+    /// State events cannot be replied to.
+    #[error("tried to reply to a state event")]
+    StateEvent,
+}
+
+/// Whether or not to enforce a [`Relation::Thread`] when sending a reply.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[allow(clippy::exhaustive_enums)]
```

Comment:
Out of curiosity, why is this line useful?

Reply:
Sorry, this crept in with https://github.com/matrix-org/matrix-rust-sdk/pull/4819/files#diff-7f49982c21dd9af0fcce15d1d1349716f42c41268483df13c1dc8086c6448d57R143, I think because I was initially copying the `ForwardThread` enum from ruma. I'll remove it.